src/AdvancedCircuitBreaker.php

```diff
@@ -0,0 +1,152 @@
+ * https://opensource.org/licenses/OSL-3.0
+ * If you did not receive a copy of the license and are unable to
+ * obtain it through the world-wide-web, please send an email
+ * to license@prestashop.com so we can send you a copy immediately.
+ *
+ * DISCLAIMER
+ *
+ * Do not edit or add to this file if you wish to upgrade PrestaShop to newer
+ * versions in the future. If you wish to customize PrestaShop for your
+ * needs please refer to http://www.prestashop.com for more information.
+ *
+ * @author    PrestaShop SA <contact@prestashop.com>
+ * @copyright 2007-2019 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ * International Registered Trademark & Property of PrestaShop SA
+ */
+
+namespace PrestaShop\CircuitBreaker;
+
+use PrestaShop\CircuitBreaker\Contracts\Client;
+use PrestaShop\CircuitBreaker\Contracts\ConfigurableCall;
+use PrestaShop\CircuitBreaker\Contracts\Storage;
+use PrestaShop\CircuitBreaker\Contracts\System;
+use PrestaShop\CircuitBreaker\Contracts\Transitioner;
+use PrestaShop\CircuitBreaker\Exceptions\UnavailableServiceException;
+
+/**
+ * This implementation of the CircuitBreaker is a bit more advanced than the SimpleCircuitBreaker,
+ * it allows you to setup your client, system or storage. You can also add an optional transitioner.
+ * And you can send requests with parameters.
+ */
+class AdvancedCircuitBreaker extends PartialCircuitBreaker implements ConfigurableCall, Transitioner
+{
+    /** @var Transitioner */
+    protected $transitioner;
+
+    /**
+     * @param System $system
+     * @param Client $client
+     * @param Storage $storage
+     * @param Transitioner|null $transitioner
+     */
+    public function __construct(System $system, Client $client, Storage $storage, Transitioner $transitioner = null)
+    {
+        parent::__construct($system, $client, $storage);
+        $this->transitioner = $transitioner;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function call($service, callable $fallback)
+    {
+        return $this->callWithParameters($service, $fallback, []);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function callWithParameters(
+        $service,
+        callable $fallback,
+        array $serviceParameters = []
+    ) {
+        $transaction = $this->initTransaction($service);
+
+        if ($this->isOpened()) {
+            if ($this->canAccessService($transaction)) {
+                $this->moveStateTo(States::HALF_OPEN_STATE, $service);
+                $this->beginTransition(
+                    Transitions::CHECKING_AVAILABILITY_TRANSITION,
+                    $service,
+                    $serviceParameters
+                );
+            }
+
+            return \call_user_func($fallback);
+        }
+
+        // Use do..while loop to allow at least one request (even when 0 retries have set
+        // in an HalfOpen state for example, or it will stay stuck in this state)
```

Comment:
true, but imho it should be validated at HalfOpenPlace level as we can consider the object invalid in this case, right?

Reply:
I agree however this could break previously implementations of the CB... I really hesitate here